ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/akkastreams/dispatcher/SignalDispatcher.scala

```diff
@@ -60,21 +60,32 @@ class SignalDispatcher private () {
 
   private def throwClosed(): Nothing = throw new IllegalStateException("SignalDispatcher is closed")
 
-  /** Closes this SignalDispatcher.
+  /** Closes this SignalDispatcher by gracefully completing the existing Source subscriptions.
     * For any downstream with pending signals, at least one such signal will be sent first.
     */
-  def shutdown(): Future[Unit] = {
+  def shutdown(): Future[Unit] = shutdownInternal(_.complete())
+
+  /** Closes this SignalDispatcher by failing the existing Source subscriptions with the provided throwable. */
+  def fail(throwable: Throwable): Future[Unit] =
+    shutdownInternal(_.fail(throwable)).transform {
+      // This throwable is expected so map to Success
+      case Failure(`throwable`) => Success(())
+      case other => other
+    }(ExecutionContext.parasitic)
+
+  private def shutdownInternal(
+      shutdownSourceQueue: SourceQueueWithComplete[_] => Unit
+  ): Future[Unit] = {
     implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.parasitic
     runningState
       .getAndSet(None)
       .fold(throw new IllegalStateException("SignalDispatcher is already closed")) { sources =>
-        sources.foreach(_.complete())
+        sources.foreach(shutdownSourceQueue)
```

Comment:
What do you think about doing this also in a Future? Can it be that this takes long (eg synchronously, and blocking a synchronized block in the DispatcherState)?

Reply:
Afais this Akka API shouldn't block. But let's wrap for safety.